### PR TITLE
SWARM-1663 Create WF Swarm fraction for Fault Tolerance impl

### PR DESCRIPTION
**Pull request by antoinesd**
Code initially developed with @mkouba in https://github.com/wildfly-swarm/microprofile-fault-tolerance-hystrix.
Code was split in cdi-extension, Fraction and TCK in testsuite. Tests were migrated from TestNG to JUnit

### CODE REVIEW

**Comment by kenfinnigan** [comment]
add to whitelist

File: fractions/microprofile/microprofile-fault-tolerance/src/main/resources/modules/org/eclipse/microprofile/faulttolerance/main/module.xml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+
+<module xmlns="urn:jboss:module:1.3" name="org.eclipse.microprofile.faulttolerance">
+  <resources>
+    <artifact name="org.eclipse.microprofile.fault-tolerance:microprofile-fault-tolerance-api:${mpft-api.version}"/>
+    <artifact name="com.netflix.hystrix:hystrix-core:${hystrix.version}"/>
+    <artifact name="org.wildfly.swarm:mp_fault_tolerance_cdi_extension:${project.version}"/>
+    <artifact name="io.reactivex:rxjava:1.2.5" />
```
Comment (mkouba) [inline fractions/microprofile/microprofile-fault-tolerance/src/main/resources/modules/org/eclipse/microprofile/faulttolerance/main/module.xml]: I think that hystrix and rxjava resources should be replaced with `com.netflix.hystrix` module dependency.

File: client-apis/mp_fault_tolerance_cdi_extension/pom.xml [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+  <properties>
+    <!-- The MicroProfile Config API version -->
+    <mpconfig.api.version>1.1</mpconfig.api.version>
+    <!-- The Wildfly Extras Version -->
+    <mpconfig.wildfly.version>1.1.2</mpconfig.wildfly.version>
+    <mpft-api.version>1.0</mpft-api.version>
+    <hystrix.version>1.5.12</hystrix.version>
+    <javax-annotation-api.version>1.3</javax-annotation-api.version>
+    <version.arquillian-weld-embedded>2.0.0.Beta5</version.arquillian-weld-embedded>
+    <version.weld>2.3.5.Final</version.weld>
```
Comment (mkouba) [inline client-apis/mp_fault_tolerance_cdi_extension/pom.xml]: `build-parent` already defines `ersion.weld-se` but you don't need a version at all, just remove `org.jboss.weld:weld-core-impl` from dependencyManagement -> the version is managed by `wildfly-parent`.

**Comment by kenfinnigan** [comment]
There's also a conflict with pom.xml that needs to be resolved

File: client-apis/mp_fault_tolerance_cdi_extension/pom.xml [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  <groupId>org.wildfly.swarm</groupId>
+  <artifactId>mp_fault_tolerance_cdi_extension</artifactId>
+
+  <name>Microprofile Fault Tolerance CDI Extension</name>
+  <description>Fault tolerance export for Microprofile</description>
+
+  <packaging>jar</packaging>
+
+  <properties>
+    <!-- The MicroProfile Config API version -->
```
Comment (kenfinnigan) [inline client-apis/mp_fault_tolerance_cdi_extension/pom.xml:37]: All these properties shouldn't be set directly in here, they should be coming from build-parent, as per https://issues.jboss.org/browse/SWARM-1627

File: client-apis/mp_fault_tolerance_cdi_extension/pom.xml [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    </dependency>-->
+
+
+    <dependency>
+      <groupId>org.jboss.arquillian.junit</groupId>
+      <artifactId>arquillian-junit-container</artifactId>
+      <scope>test</scope>
+    </dependency>
+
+    <dependency>
```
Comment (kenfinnigan) [inline client-apis/mp_fault_tolerance_cdi_extension/pom.xml:94]: Why is weld-core-impl a provided and compile dependency?

File: client-apis/mp_fault_tolerance_cdi_extension/pom.xml [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+      <groupId>org.jboss.weld</groupId>
+      <artifactId>weld-core-impl</artifactId>
+    </dependency>
+
+
+    <dependency>
```
Comment (kenfinnigan) [inline client-apis/mp_fault_tolerance_cdi_extension/pom.xml]: Why do we need arquillian-weld-embedded for a unit test?

File: fractions/microprofile/microprofile-fault-tolerance/pom.xml [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+  <name>Eclipse MicroProfile Fault Tolerance</name>
+  <description>WildFly: Swarm Fraction for Eclipse MicroProfile Fault Tolerance</description>
+
+  <packaging>jar</packaging>
+
+  <properties>
+    <swarm.fraction.internal>false</swarm.fraction.internal>
+    <swarm.fraction.stability>experimental</swarm.fraction.stability>
+    <swarm.fraction.tags>EclipseMicroProfile,MicroServices,FaultTolerance</swarm.fraction.tags>
+    <!-- The MicroProfile Config API version -->
```
Comment (kenfinnigan) [inline fractions/microprofile/microprofile-fault-tolerance/pom.xml:39]: These versions should be coming from build-parent

File: pom.xml [hunk 7]
```diff
@@ -1125,6 +1138,7 @@
       <modules>
         <module>client-apis/jaxrs-client-api</module>
         <module>client-apis/mp_metrics_cdi_extension</module>
+        <module>client-apis/mp_fault_tolerance_cdi_extension</module>
```
Comment (kenfinnigan) [inline pom.xml]: The CDI extension doesn't need to be referenced twice, it's separate profiles that are combined

File: testsuite/microprofile-tcks/pom.xml [hunk 8]
```diff
@@ -43,6 +43,8 @@
         <version.mpjwt.auth>1.0</version.mpjwt.auth>
         <!-- The MicroProfile Heath Check API version -->
         <version.mphealth>1.0</version.mphealth>
+        <!-- The MicroProfile Fault Tolerance API version -->
```
Comment (kenfinnigan) [inline testsuite/microprofile-tcks/pom.xml:47]: These versions shouldn't live in here

File: fractions/microprofile/microprofile-fault-tolerance/src/main/resources/modules/org/eclipse/microprofile/faulttolerance/main/module.xml (hunk 1 above)
Comment (kenfinnigan) [reply fractions/microprofile/microprofile-fault-tolerance/src/main/resources/modules/org/eclipse/microprofile/faulttolerance/main/module.xml]: this custom module seems wrong.

If it's an Eclipse MP FT module then it shouldn't have rxjava, hystrix, or our cdi extension as dependencies.

Is this trying to be a deployment module? I think there's a better way to do that

**Review by kenfinnigan — CHANGES_REQUESTED** [review]

**Comment by mkouba** [comment]
@kenfinnigan @antoinesd I would fix all the version location-related problems in a different PR (for all MP fractions at once).

**Comment by kenfinnigan** [comment]
There is a JIRA for that, so makes sense.

Main question would be whether this PR needs to come after that one, or that one fixes all these instances as well?

**Comment by mkouba** [comment]
+1 for merging this PR first. Afterwards, I'd like to come with a PR for SWARM-1644 and then a PR for SWARM-1627.

**Comment by kenfinnigan** [comment]
I'm fine with that.

@antoinesd, can you fix the conflict with pom.xml and any other items that were raised unrelated to where versions are located, and we will get it merged

**Comment by antoinesd** [comment]
@mkouba, @kenfinnigan I'm still working on integrating history from previous repository. I still have issue with the final rebase.

File: client-apis/mp_fault_tolerance_cdi_extension/pom.xml (hunk 2 above)
Comment (antoinesd) [reply client-apis/mp_fault_tolerance_cdi_extension/pom.xml]: Replacing `org.jboss.weld:weld-core-impl` by `org.wildfly:wildfly-weld` and removing `dependencyManagement`

**Comment by antoinesd** [comment]
Rebased on master, made requested modification and added History of the project (tricky part)

File: fractions/microprofile/microprofile-fault-tolerance/src/main/resources/modules/org/eclipse/microprofile/faulttolerance/main/module.xml [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+  ~
+  ~ Unless required by applicable law or agreed to in writing, software
+  ~ distributed under the License is distributed on an "AS IS" BASIS,
+  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+
+<module xmlns="urn:jboss:module:1.3" name="org.eclipse.microprofile.faulttolerance">
+
```
Comment (mkouba) [inline fractions/microprofile/microprofile-fault-tolerance/src/main/resources/modules/org/eclipse/microprofile/faulttolerance/main/module.xml:18]: Empty? How do you actually "enable" MP FT per deployment?

**Review by kenfinnigan — APPROVED** [review]

**Pull request merged**